### DISCUSSION TITLE
Error: Invalid function: static-if

### DISCUSSION THREAD

**Discussion by YueRen** [minad/corfu#482] [Q&A] [answered]
I currently get the following emacs whenever I start my emacs (Ubuntu 22.04, emacs v29.4 installed via snap):

```
⛔ Error (use-package): corfu/:init: Invalid function: static-if
⛔ Error (use-package): corfu/:catch: Invalid function: static-if
⛔ Error (use-package): corfu-prescient/:init: Invalid function: static-if
⛔ Error (use-package): corfu-prescient/:catch: Invalid function: static-if
```

I also get similar error messages in various other locations, e.g., `Error in pre-command-hook (vertico--prepare): (invalid-function static-if)` when trying to use tab-completion in the minibuffer.

Is this purely a problem on my system or a general problem with emacs v29.4?  The following github issue sounds as if one needs emacs v30 for `static-if`:
https://github.com/doomemacs/doomemacs/issues/7939

**Comment by minad** [comment, marked as answer]
Reinstalling Compat, Vertico and Corfu and restarting Emacs should fix this problem.

**Reply by YueRen** [reply]
Thanks for the quick suggestion, I've tried `M-x package-reinstall` + `compat`, `vertico`, `corfu`, but to no avail.

However, I see what was causing the issue.  `M-x package-list-packages` lists `compat` as version `30.0.0.0`.  However, 
* my `~/.emacs.d/elpa/compat-30.0.0.0/` folder naturally had version `30.0.0.0`,
* my `~/.emacs.d/straight/repos/compat` folder only had version `29.1.4.5`.

Somehow, `package-list-packages` listed the version in `~/.emacs.d/elpa/compat-30.0.0.0/`, but my emacs loaded the version in `~/.emacs.d/straight/repos/compat`.  Not sure whether it was a good idea, but I've manually pulled the remote master branch in `~/.emacs.d/straight/repos/compat` and the error is now gone.

**Comment by bilus** [comment]
What worked for me is pinning compat in `packages.el`:

```elisp
(package! compat :pin "9a234d0")
```

The relevant commits for the `compat` version:

```
        9a234d0 compat.texi: Fix replace-string-in-region documentation (Fix #52)
        e9203e1 compat.texi: Fix references to Emacs 30.1 in Support section (#49)
```
